pkg/kubepkg/kubepkg.go

```diff
@@ -191,169 +241,135 @@ func buildPackage(pkg, arch string, packageDef *PackageDefinition) error {
+	bc.KubernetesVersion = strings.Replace(bc.KubernetesVersion, "+", "-", 1)
 
-	c.Version, err = getPackageVersion(packageDef)
+	bc.Version, err = getPackageVersion(packageDef)
 	if err != nil {
 		return errors.Wrap(err, "getting package version")
 	}
 
-	logrus.Infof("%s package version: %s", c.Name, c.Version)
+	logrus.Infof("%s package version: %s", bc.Name, bc.Version)
 
-	c.KubeletCNIVersion = minimumCNIVersion
+	bc.KubeletCNIVersion = minimumCNIVersion
 
-	c.Dependencies, err = GetKubeadmDependencies(packageDef)
+	bc.Dependencies, err = GetKubeadmDependencies(packageDef)
 	if err != nil {
 		return errors.Wrap(err, "getting kubeadm dependencies")
 	}
 
-	c.KubeadmKubeletConfigFile = kubeadmConf
+	bc.KubeadmKubeletConfigFile = kubeadmConf
 
-	if c.Arch == "arm" {
-		c.DebArch = "armhf"
-	} else if c.Arch == "ppc64le" {
-		c.DebArch = "ppc64el"
-	} else {
-		c.DebArch = c.Arch
-	}
+	bc.BuildArch = getBuildArch(bc.GoArch, bc.Type)
 
-	c.CNIDownloadLink, err = getCNIDownloadLink(packageDef, c.Arch)
+	bc.CNIDownloadLink, err = getCNIDownloadLink(packageDef, bc.BuildArch)
 	if err != nil {
 		return errors.Wrap(err, "getting CNI download link")
 	}
 
-	logrus.Infof("Building %s package for %s/%s architecture...", c.Package, c.Arch, c.DebArch)
-	return c.run()
+	logrus.Infof("Building %s package for %s/%s architecture...", bc.Package, bc.GoArch, bc.BuildArch)
+	return bc.run()
 }
 
-func (c cfg) run() error {
-	var w []work
-
-	// nolint
-	// TODO: Get package directory for any version once package definitions are broken out
-	srcdir := filepath.Join(latestPackagesDir, c.Package)
-	dstdir, err := ioutil.TempDir(os.TempDir(), "debs")
+func (bc *buildConfig) run() error {
+	workspaceInfo, err := os.Stat(bc.workspace)
 	if err != nil {
 		return err
 	}
 
-	if !*keepTmp {
-		defer os.RemoveAll(dstdir)
-	}
-
-	if err := filepath.Walk(srcdir, func(srcfile string, f os.FileInfo, err error) error {
-		if err != nil {
-			return err
-		}
-		dstfile := filepath.Join(dstdir, srcfile[len(srcdir):])
-		if dstfile == dstdir {
-			return nil
-		}
-		if f.IsDir() {
-			return os.Mkdir(dstfile, f.Mode())
-		}
-		t, err := template.
-			New("").
-			Funcs(builtins).
-			Option("missingkey=error").
-			ParseFiles(srcfile)
-		if err != nil {
-			return err
-		}
-		w = append(w, work{
-			src:  srcfile,
-			dst:  dstfile,
-			t:    t.Templates()[0],
-			info: f,
-		})
+	specDir := filepath.Join(bc.workspace, string(bc.Channel), bc.Package)
+	specDirWithArch := filepath.Join(specDir, bc.GoArch)
 
-		return nil
-	}); err != nil {
+	err = os.MkdirAll(specDirWithArch, workspaceInfo.Mode())
+	if err != nil {
 		return err
 	}
 
-	for _, w := range w {
-		if err := func() error {
-			f, err := os.OpenFile(w.dst, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0)
-			if err != nil {
-				return err
-			}
-			defer f.Close()
+	//nolint:godox
+	// TODO: keepTmp/cleanup needs to defined in kubepkg root
+	if !bc.specOnly {
+		defer os.RemoveAll(specDirWithArch)
```

Comment:
This will also remove the `bc.workspace`, which should be fine if we implicitly create it as mentioned in the comment above.

Reply:
I'll fix this up once I add the cleanup/don't cleanup(?) flag.